File: server/python/tests/test_all.py
#!/usr/bin/env python3
"""
Quick test script to validate all Python code samples.
This script checks:
1. Python syntax validity
2. Import dependencies
3. Basic structure validation
"""

import ast
import importlib.util
import sys
from pathlib import Path
from typing import List, Tuple

# Colors for output
GREEN = '\033[92m'
RED = '\033[91m'
YELLOW = '\033[93m'
BLUE = '\033[94m'
RESET = '\033[0m'

def check_syntax(file_path: Path) -> Tuple[bool, str]:
    """Check if a Python file has valid syntax."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            source = f.read()
        ast.parse(source, filename=str(file_path))
        return True, "✓ Syntax valid"
    except SyntaxError as e:
        return False, f"✗ Syntax error: {e.msg} at line {e.lineno}"
    except Exception as e:
        return False, f"✗ Error reading file: {e}"

def check_imports(file_path: Path) -> Tuple[bool, str]:
    """Check if required imports can be resolved."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            source = f.read()
        
        tree = ast.parse(source)
        imports = []
        for node in ast.walk(tree):
            if isinstance(node, ast.Import):
                for alias in node.names:
                    imports.append(alias.name)
            elif isinstance(node, ast.ImportFrom):
                if node.module:
                    imports.append(node.module)
        
        # Check if critical imports exist
        # Note: trycourier package is imported as 'courier'
        required_imports = ['courier', 'dotenv']
        missing = []
        for imp in required_imports:
            if any(imp in i for i in imports):
                try:
                    if imp == 'courier':
                        import courier  # This comes from trycourier package
                    elif imp == 'dotenv':
                        import dotenv  # This comes from python-dotenv package
                except ImportError:
                    # Map import names to package names for error messages
                    package_name = 'trycourier' if imp == 'courier' else 'python-dotenv'
                    missing.append(package_name)
        
        if missing:
            return False, f"✗ Missing dependencies: {', '.join(missing)}"
        return True, "✓ Imports valid"
    except Exception as e:
        return False, f"✗ Error checking imports: {e}"

def check_structure(file_path: Path) -> Tuple[bool, str]:
    """Check basic structure - Courier client initialization."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            source = f.read()
        
        # Check for key patterns
        checks = []
        if 'from courier import Courier' in source or 'import courier' in source:
            checks.append("Courier import")
        if 'Courier(' in source or 'courier.Courier(' in source:
            checks.append("Courier client initialization")
        if 'load_dotenv' in source:
            checks.append("Environment loading")
        
        if len(checks) >= 2:
            return True, f"✓ Structure valid ({', '.join(checks)})"
        else:
            return False, f"✗ Missing key components: {checks}"
    except Exception as e:
        return False, f"✗ Error checking structure: {e}"

def main():
    """Main test runner."""
    # Test files are in tests/ directory, but Python samples are one level up
    script_dir = Path(__file__).parent.parent
    python_files = sorted(script_dir.glob("*.py"))
    
    # Exclude this test script and any other test files
    python_files = [f for f in python_files if f.name != "test_all.py" and not f.name.startswith("test_")]
    
    if not python_files:
        print(f"{RED}No Python files found to test{RESET}")
        return 1
    
    print(f"{BLUE}Testing {len(python_files)} Python sample files...{RESET}\n")
    
    results = []
    for file_path in python_files:
        print(f"{BLUE}Testing: {file_path.name}{RESET}")
        
        syntax_ok, syntax_msg = check_syntax(file_path)
        imports_ok, imports_msg = check_imports(file_path)
        structure_ok, structure_msg = check_structure(file_path)
        
        all_ok = syntax_ok and imports_ok and structure_ok
        
        if syntax_ok:
            print(f"  {GREEN}{syntax_msg}{RESET}")
        else:
            print(f"  {RED}{syntax_msg}{RESET}")
        
        if imports_ok:
            print(f"  {GREEN}{imports_msg}{RESET}")
        else:
            print(f"  {YELLOW}{imports_msg}{RESET} (install with: pip install -r requirements.txt)")
        
        if structure_ok:
            print(f"  {GREEN}{structure_msg}{RESET}")
        else:
            print(f"  {YELLOW}{structure_msg}{RESET}")
        
        results.append((file_path.name, all_ok))
        print()
    
    # Summary
    print(f"{BLUE}{'='*60}{RESET}")
    print(f"{BLUE}Summary:{RESET}\n")
    
    passed = sum(1 for _, ok in results if ok)
    total = len(results)
    
    for name, ok in results:
        status = f"{GREEN}✓ PASS{RESET}" if ok else f"{YELLOW}⚠ PARTIAL{RESET}"
        print(f"  {status} - {name}")
    
    print(f"\n{BLUE}Results: {passed}/{total} files passed all checks{RESET}")
    
    if passed == total:
        print(f"{GREEN}All files are valid! ✓{RESET}")
        return 0
    else:
        print(f"{YELLOW}Some files need attention (likely missing dependencies){RESET}")
        print(f"{YELLOW}Run: pip install -r requirements.txt{RESET}")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: server/node/tests/test_all.js
#!/usr/bin/env node
/**
 * Test script to validate all Node.js code samples.
 * This script checks:
 * 1. JavaScript syntax validity
 * 2. Import dependencies
 * 3. Basic structure validation
 */

import { readdir, readFile } from 'fs/promises';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = createRequire(import.meta.url);

// Colors for output
const GREEN = '\033[92m';
const RED = '\033[91m';
const YELLOW = '\033[93m';
const BLUE = '\033[94m';
const RESET = '\033[0m';

async function checkSyntax(filePath) {
    try {
        // Try to require/import the file to check syntax
        // Note: We can't easily check ES module syntax without executing
        // So we'll just check if file is readable
        await readFile(filePath, 'utf-8');
        return { ok: true, message: '✓ Syntax valid (file readable)' };
    } catch (error) {
        return { ok: false, message: `✗ Error reading file: ${error.message}` };
    }
}

async function checkImports(filePath) {
    try {
        const content = await readFile(filePath, 'utf-8');
        
        // Check for required imports
        const hasCourier = content.includes('@trycourier/courier') || content.includes("from '@trycourier/courier'");
        const hasDotenv = content.includes('dotenv') || content.includes("from 'dotenv'");
        
        const missing = [];
        if (!hasCourier) {
            missing.push('@trycourier/courier');
        }
        if (!hasDotenv) {
            missing.push('dotenv');
        }
        
        if (missing.length > 0) {
            return { ok: false, message: `✗ Missing imports: ${missing.join(', ')}` };
        }
        
        // Try to check if packages are installed
        try {
            require.resolve('@trycourier/courier');
            require.resolve('dotenv');
            return { ok: true, message: '✓ Imports valid' };
        } catch (err) {
            return { ok: false, message: '✗ Dependencies not installed (run: npm install)' };
        }
    } catch (error) {
        return { ok: false, message: `✗ Error checking imports: ${error.message}` };
    }
}

async function checkStructure(filePath) {
    try {
        const content = await readFile(filePath, 'utf-8');
        
        const checks = [];
        if (content.includes('Courier') || content.includes('new Courier')) {
            checks.push('Courier client');
        }
        if (content.includes('dotenv') || content.includes('dotenv.config')) {
            checks.push('Environment loading');
        }
        if (content.includes('COURIER_API_KEY')) {
            checks.push('API key reference');
        }
        
        if (checks.length >= 2) {
            return { ok: true, message: `✓ Structure valid (${checks.join(', ')})` };
        } else {
            return { ok: false, message: `✗ Missing key components` };
        }
    } catch (error) {
        return { ok: false, message: `✗ Error checking structure: ${error.message}` };
    }
}

async function main() {
    const parentDir = join(__dirname, '..');
    const files = await readdir(parentDir);
    
    // Filter for .js files, exclude test files
    const jsFiles = files
        .filter(f => f.endsWith('.js') && !f.startsWith('test_'))
        .sort();
    
    if (jsFiles.length === 0) {
        console.error(`${RED}No JavaScript files found to test${RESET}`);
        process.exit(1);
    }
    
    console.log(`${BLUE}Testing ${jsFiles.length} Node.js sample files...${RESET}\n`);
    
    const results = [];
    for (const fileName of jsFiles) {
        const filePath = join(parentDir, fileName);
        console.log(`${BLUE}Testing: ${fileName}${RESET}`);
        
        const syntax = await checkSyntax(filePath);
        const imports = await checkImports(filePath);
        const structure = await checkStructure(filePath);
        
        if (syntax.ok) {
            console.log(`  ${GREEN}${syntax.message}${RESET}`);
        } else {
            console.log(`  ${RED}${syntax.message}${RESET}`);
        }
        
        if (imports.ok) {
            console.log(`  ${GREEN}${imports.message}${RESET}`);
        } else {
            console.log(`  ${YELLOW}${imports.message}${RESET}`);
        }
        
        if (structure.ok) {
            console.log(`  ${GREEN}${structure.message}${RESET}`);
        } else {
            console.log(`  ${YELLOW}${structure.message}${RESET}`);
        }
        
        const allOk = syntax.ok && imports.ok && structure.ok;
        results.push({ name: fileName, ok: allOk });
        console.log('');
    }
    
    // Summary
    console.log(`${BLUE}${'='.repeat(60)}${RESET}`);
    console.log(`${BLUE}Summary:${RESET}\n`);
    
    const passed = results.filter(r => r.ok).length;
    const total = results.length;
    
    for (const result of results) {
        const status = result.ok ? `${GREEN}✓ PASS${RESET}` : `${YELLOW}⚠ PARTIAL${RESET}`;
        console.log(`  ${status} - ${result.name}`);
    }
    
    console.log(`\n${BLUE}Results: ${passed}/${total} files passed all checks${RESET}`);
    
    if (passed === total) {
        console.log(`${GREEN}All files are valid! ✓${RESET}`);
        process.exit(0);
    } else {
        console.log(`${YELLOW}Some files need attention (likely missing dependencies)${RESET}`);
        console.log(`${YELLOW}Run: npm install${RESET}`);
        process.exit(1);
    }
}

main().catch(error => {
    console.error(`${RED}Error: ${error.message}${RESET}`);
    process.exit(1);
});

